### ISSUE TITLE
Spurious type error in `except` branch after an assignment from an `await` that throws

### ISSUE THREAD

**Issue by bcmills** [python/mypy#18074]
**Bug Report**

If a loop is guarded with an `x is not None` condition, and the body of the loop contains a `try`...`except` that assigns to `x`, `mypy` erroneously concludes that `x` might be `None` in the `except` branch.

**To Reproduce**

(https://mypy-play.net/?mypy=latest&python=3.12&gist=000b86de168d7bca72f3ff58de4121cb)

```python
import asyncio
from collections.abc import Awaitable
from typing import Optional

R = Awaitable[Optional['R']]

async def boom() -> R:
	raise ValueError("Boom!")

async def main() -> None:
	x: Optional[R] = boom()
	y: R
	while x is not None:
		try:
			x = await x
		except:
			y = x
			break

asyncio.run(main())
```

**Expected Behavior**

The program is well-typed.

**Actual Behavior**

```
main.py:17: error: Incompatible types in assignment (expression has type "R | None", variable has type "R")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

See Playground details.

**Comment by A5rocks** [comment]
This is happening because mypy doesn't have special casing around when a try block when or won't throw. This means your example could as well be:
```py
import asyncio
from collections.abc import Awaitable
from typing import Optional

R = Awaitable[Optional['R']]

async def boom() -> R:
	raise ValueError("Boom!")

async def main() -> None:
	x: Optional[R] = boom()
	y: R
	while x is not None:
		try:
			x = await x
			assert False
		except:
			y = x
			break

asyncio.run(main())
```

which obviously can have `x` be `None`. I think this would probably be complicated to actually solve, for not much benefit (I can't imagine this pattern is common?).
